docs/kube-api-custom-ocp-versions.md

```diff
@@ -0,0 +1,94 @@
+* [OPENSHIFT_VERSIONS environment variable](https://github.com/openshift/assisted-service/blob/master/hack/generate.sh#L61)
+
+### [ClusterImageSet](https://github.com/openshift/hive/blob/master/apis/hive/v1/clusterimageset_types.go)
+
+The ClusterImageSet is used for referencing to a OpenShift release image.
+So available versions are represented in an Hive cluster by defined ClusterImageSet resources.
+To use a specific release image, it should be defined in the [ClusterDeployment CRD](https://github.com/openshift/hive/blob/master/docs/using-hive.md#clusterdeployment) in either of the following manners:
+* As a reference to the ClusterImageSet in ```spec.provisioning.imageSetRef``` property.
+* Explicitly as a URL in ```spec.provisioning.releaseImage``` property.
+
+An example of a ClusterImageSet:
+```
+apiVersion: hive.openshift.io/v1
+kind: ClusterImageSet
+metadata:
+  name: openshift-v4.7.0
+spec:
+  releaseImage: quay.io/openshift-release-dev/ocp-release:4.7.0-x86_64
+```
+
+### [OPENSHIFT_VERSIONS](https://github.com/openshift/assisted-service/blob/master/Makefile#L36)
+
+The OPENSHIFT_VERSIONS environment variable is a map of OpenShift version (y-stream) to related properties as release version/image and RHCOS version/image: see [default_ocp_versions.json](https://github.com/openshift/assisted-service/blob/master/default_ocp_versions.json)
+
+Defining this var prior to starting the service would override all versions supported by the service. I.e. by default, [default_ocp_versions.json](https://github.com/openshift/assisted-service/blob/master/default_ocp_versions.json) is used for setting the supported versions. Defining OPENSHIFT_VERSIONS with a similar json format should override these.
+
+## Flow
+
+The flow of adding a new version is a follows:
+* OPENSHIFT_VERSIONS is defined with a map of "x.y" -> ```rhcos_image``` and ```rhcos_version```.
+* Upon starting the service, the relevant host [boot-files](https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L16) are uploaded to S3/File storage.
+* ClusterImageSet with a new ```releaseImage``` is deployed to Hive cluster.
+* The name of the ClusterImageSet is referenced in the ClusterDeployment (```spec.provisioning.imageSetRef``` property).
+* Upon deploying the ClusterDeployment:
+  * ClusterImageSet is fetched according to ```imageSetRef```.
+  * ```releaseImage``` is read from the ClusterImageSet.
+  * The ```release_version``` is extracted from the release image (using ```oc adm release info```) - in major.minor.patch format.
+  * According to the major.minor as key, the relevant entry is updated in the [versions cache](https://github.com/openshift/assisted-service/blob/master/internal/versions/versions.go). I.e. ```release_version``` and ```display_name``` properties.
+* Finally, a new cluster is created with the newly added [openshift_version](https://github.com/openshift/assisted-service/blob/master/swagger.yaml#L4106). 
```

Comment:
if I remember correctly there is another way to mark specific lines with blame or something like that, this link will not be relevant if someone adds or removes lines from the swagger.